File: repository/src/main/java/com/sreMake/repository/user/impl/UserRepositoryImpl.java
package com.sreMake.repository.user.impl;


import com.sreMake.model.Tables;
import com.sreMake.model.user.User;
import com.sreMake.model.user.dto.UserSearchInput;
import com.sreMake.repository.user.UserRepository;
import org.babyfish.jimmer.Page;
import org.babyfish.jimmer.spring.repo.PageParam;
import org.babyfish.jimmer.spring.repo.support.AbstractJavaRepository;
import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.ast.LikeMode;
import org.springframework.stereotype.Repository;

import java.util.Objects;

@Repository
public class UserRepositoryImpl extends AbstractJavaRepository<User, Long> implements UserRepository {


    public UserRepositoryImpl(JSqlClient sql) {
        super(sql);
    }

    @Override
    public Page<User> findPage(PageParam pageParam, UserSearchInput params) {
        if (Objects.isNull(params)) {
            return this.findPage(pageParam);
        } else {
            return this.sql.createQuery(Tables.USER_TABLE).where(
                            Tables.USER_TABLE.id().eqIf(params.getId()),
                            Tables.USER_TABLE.username().likeIf(params.getUsername(), LikeMode.ANYWHERE),
                            Tables.USER_TABLE.status().eqIf(params.getStatus()),
                            Tables.USER_TABLE.phone().likeIf(params.getPhone(), LikeMode.ANYWHERE),
                            Tables.USER_TABLE.createAt().betweenIf(params.getCreateAtStart(), params.getCreateAtEnd()),
                            Tables.USER_TABLE.updateAt().betweenIf(params.getUpdateAtStart(), params.getUpdateAtEnd())

                    ).orderBy(Tables.USER_TABLE.id().desc())
                    .select(Tables.USER_TABLE)
                    .fetchPage(pageParam.getIndex(), pageParam.getSize());
        }

    }

    @Override
    public User findByUsername(String username) {
        return this.sql.createQuery(Tables.USER_TABLE).where(
                        Tables.USER_TABLE.username().eq(username),
                        Tables.USER_TABLE.status().eq(User.Status.NORMAL)
                ).select(Tables.USER_TABLE)
                .fetchOne();
    }
}

File: repository/src/main/kotlin/com/SReMake/repository/user/impl/UserRepositoryImpl.kt
package com.SReMake.repository.user.impl

import com.SReMake.model.user.*
import com.SReMake.model.user.dto.UserSearchInput
import com.SReMake.repository.user.UserRepository
import org.babyfish.jimmer.spring.repo.PageParam
import org.babyfish.jimmer.spring.repo.support.AbstractKotlinRepository
import org.babyfish.jimmer.sql.ast.LikeMode
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.ast.expression.*
import org.springframework.stereotype.Repository

@Repository
class UserRepositoryImpl(sql: KSqlClient) : UserRepository, AbstractKotlinRepository<User, Long>(sql) {
    override fun findPage(pageParam: PageParam, params: UserSearchInput?) =
        params?.let {
            this.sql.createQuery(User::class) {
                where(
                    table.id.`eq?`(params.id),
                    table.username.`ilike?`(params.username, LikeMode.ANYWHERE),
                    table.status.`eq?`(params.status),
                    table.phone.`ilike?`(params.phone, LikeMode.ANYWHERE),
                    table.createAt.`between?`(params.createAtStart, params.createAtEnd),
                    table.updateAt.`between?`(params.updateAtStart, params.updateAtEnd)
                )
                orderBy(table.id.desc())
                select(table)
            }.fetchPage(pageParam.index, pageParam.size)
        } ?: this.findPage(pageParam);


    override fun findByUsername(username: String) =
        this.sql.createQuery(User::class) {
            where(
                table.username.eq(username),
                table.status.eq(Status.Normal)
            )
            select(table)
        }.fetchOne()


}